File: sdks/typescript/packages/sdk/src/scenario.ts
/**
 * Define a named scenario.
 *
 * A scenario's `up` is free-form async code (loops, conditionals, real
 * API calls) that provisions an isolated environment and returns the
 * `auth`/`teardown` a test run needs. An omitted `down` is a no-op. Register
 * scenarios with `createHandler({ ..., scenarios: [defineScenario({...})] })`.
 *
 * @example
 * ```ts
 * defineScenario({
 *   name: 'single-user',
 *   description: 'One verified user in a fresh org',
 *   up: async ({ testRunId }) => {
 *     const email = uniqueEmail(testRunId)
 *     const user = await db.user.create({ email })
 *     return {
 *       auth: { headers: { Authorization: `Bearer ${await mintToken(user)}` } },
 *       teardown: { userId: user.id },
 *     }
 *   },
 *   down: async ({ teardown }) => {
 *     await db.user.delete({ id: teardown.userId })
 *   },
 * })
 * ```
 */
import type { ScenarioDefinition } from './types'

export function defineScenario(definition: ScenarioDefinition): ScenarioDefinition {
  if (typeof definition.name !== 'string' || definition.name.length === 0) {
    throw new Error('Scenario "name" must be a non-empty string')
  }
  if (typeof definition.description !== 'string') {
    throw new Error('Scenario "description" must be a string')
  }
  if (typeof definition.up !== 'function') {
    throw new Error('Scenario "up" must be a function')
  }
  if (definition.down !== undefined && typeof definition.down !== 'function') {
    throw new Error('Scenario "down" must be a function if provided')
  }
  return definition
}

File: sdks/java/autonoma-sdk/src/main/java/ai/autonoma/sdk/Scenario.java
package ai.autonoma.sdk;

import ai.autonoma.sdk.types.ScenarioDefinition;
import ai.autonoma.sdk.types.ScenarioDownContext;
import ai.autonoma.sdk.types.ScenarioUpContext;
import ai.autonoma.sdk.types.ScenarioUpResult;

/**
 * Builds a {@link ScenarioDefinition} from lambdas.
 *
 * <p>A scenario's {@code up} is free-form code (loops, conditionals, real API
 * calls) that provisions an isolated environment and returns the
 * {@code auth}/{@code teardown} a test run needs. An omitted {@code down} is a
 * no-op. Register scenarios with
 * {@code new HandlerConfig(shared, signing, List.of(Scenario.define(...)))}.
 *
 * <pre>{@code
 * Scenario.define(
 *     "single-user",
 *     "One verified user in a fresh org",
 *     ctx -> new ScenarioUpResult(
 *         AuthResult.ofHeaders(Map.of("Authorization", "Bearer " + mintToken(ctx.testRunId()))),
 *         Map.of("userId", userId)),
 *     ctx -> deleteUser((String) ctx.teardown().get("userId")));
 * }</pre>
 */
public final class Scenario {

    private Scenario() {}

    /** Free-form provisioning function. */
    @FunctionalInterface
    public interface UpFn {
        ScenarioUpResult up(ScenarioUpContext ctx) throws Exception;
    }

    /** Optional teardown function. */
    @FunctionalInterface
    public interface DownFn {
        void down(ScenarioDownContext ctx) throws Exception;
    }

    /** Define a scenario with no teardown. */
    public static ScenarioDefinition define(String name, String description, UpFn up) {
        return define(name, description, up, null);
    }

    /** Define a scenario with a teardown. A null {@code down} is a no-op. */
    public static ScenarioDefinition define(String name, String description, UpFn up, DownFn down) {
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("Scenario \"name\" must be a non-empty string");
        }
        if (description == null) {
            throw new IllegalArgumentException("Scenario \"description\" must be a string");
        }
        if (up == null) {
            throw new IllegalArgumentException("Scenario \"up\" must be a function");
        }
        return new ScenarioDefinition() {
            @Override
            public String name() { return name; }

            @Override
            public String description() { return description; }

            @Override
            public ScenarioUpResult up(ScenarioUpContext ctx) throws Exception {
                return up.up(ctx);
            }

            @Override
            public void down(ScenarioDownContext ctx) throws Exception {
                if (down != null) down.down(ctx);
            }
        };
    }
}
